### PR TITLE
DEV-34299; fix: trigger accurate input events to make ckeditor replacements work

### PR DESCRIPTION
**Pull request by abhijeetnarvekar**
CKEditor 5 replacement issue

Info:
CKEditor is usually attached to a ContentEditable div or TextArea referred to as **container** in below writeup

How replacements used to work before:

- the replacement was done inside the text area by modifying the text nodes inside the container
- CKEditor when fetching the content pointed directly to the container
- So any changes made to the container were valid changes for CKEditor

As web technology has improved with React based applications and principles, CKEditor now maintains a state

- When CKEditor5 is initialized on a container, it takes over the container and replaces it with its overlay editing area
- This overlayed editing area is managed based on internal state.
  - If you make DOM changes to the overlay editing area programmatically(via code), the state maintained inside the editor conflicts, and editor reapplies content from the state to the overlay, making those changes irrelevant.
  - This happens in a millisecond and is not noticeable to human eye,
  - So we see replacements made by Sidebar don't stay

How do we solve this

- As we know we can certainly type manually(using keyboard) in the editing area and those changes are honored by the editor
- We need to understand difference between programmatically editing the DOM and manually typing into the editor using keyboard
  - When we press a key on keyboard certain events are triggered
  - We are here interested in **input** events
  - These input events propagate from the text node(where human types) up in the DOM.  Called as bubbles in the JS language
  - Usually there are always pair of events, before and after. Example "beforeinput" and "input"
- So, if we trigger these events before and after we make change to DOM, it triggers a change to be encountered by editor state.
- The state then reads the event information, which needs to be accurate, with offsets and data and updates itself
- This eventually also updates the DOM and our changes can now persist!

Tests:
37 Skipped CKEdior 5 Tests are now enabled, which takes total from 508 to 545

Known Issues:
Replacement will not work on CKEditor on Safari browser

Reason:
Safari does not implement the constructor for targetRanges as other browsers

Chromium implementation https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/core/events/input_event.cc;l=110?q=getTargetRanges:

WebKit implementation: https://github.com/WebKit/WebKit/blob/5a3863e89df55cd46e8dc1cdb134eb2596a454dc/Source/WebCore/dom/InputEvent.cpp#L56

By comparison you can see that `m_targetRanges` is not initialized in WebKit

There is an open PR to implement the same https://github.com/WebKit/WebKit/pull/19346

Compatibility

| Chrome |  FireFox| Safari | Edge(Chromium) |
|---------|---------|-------|------------------|
| ✅ | ✅ | ❌  | ✅ |

### CODE REVIEW

File: src/adapters/AbstractRichtextEditorAdapter.ts [hunk 1]
```diff
@@ -202,18 +202,68 @@ export abstract class AbstractRichtextEditorAdapter implements AdapterInterface
...
+          node: completeRange.startContainer,
+          eventType: 'input',
+          replacement: match.originalMatch.replacement,
+          startOffset,
+          endOffset,
+          disableSimulation: this.config.disableInputEventSimulation,
+        });
 
         removeEmptyTextNodesIfNeeded(tail);
+        removeEmptyTextNodesIfNeeded(completeRange);
```
Comment (pcdeshmukh) [inline src/adapters/AbstractRichtextEditorAdapter.ts:235]: May this is not needed.

File: src/adapters/AbstractRichtextEditorAdapter.ts [hunk 2]
```diff
@@ -202,18 +202,68 @@ export abstract class AbstractRichtextEditorAdapter implements AdapterInterface
       if (rangeLength > 1) {
         const tail = this.createRange(match.range[0] + 1, rangeLength - 1, textDomMapping);
         const head = this.createRange(match.range[0], 1, textDomMapping);
+        const completeRange = this.createRange(match.range[0], rangeLength, textDomMapping);
+
+        const { startOffset, endOffset } = completeRange;
+        console.log(`Input Event Range: ${startOffset} ${endOffset}`);
```
Comment (pcdeshmukh) [inline src/adapters/AbstractRichtextEditorAdapter.ts:208]: Remove extra console

File: src/utils/utils.ts [hunk 3]
```diff
@@ -46,48 +46,38 @@ export function isIFrame(el: Element): el is HTMLIFrameElement {
...
+  startOffset: number;
+  endOffset: number;
+  replacement: string;
+  disableSimulation?: boolean;
+};
+
+export function simulateInputEvent(props: SimulateInputEventProps) {
+  const { node, eventType, startOffset, endOffset, replacement, disableSimulation } = props;
+  if (disableSimulation) {
+    console.log('Input Event simulation is disabled for adapter');
```
Comment (pcdeshmukh) [inline src/utils/utils.ts:61]: Remove console

File: src/utils/utils.ts [hunk 4]
```diff
@@ -46,48 +46,38 @@ export function isIFrame(el: Element): el is HTMLIFrameElement {
...
     return;
   }
-
-  const textNodeContent = textNode.textContent || '';
+  console.log('Simulating input event for type:', props);
```
Comment (pcdeshmukh) [inline src/utils/utils.ts:64]: Remove console

**Review by pcdeshmukh — CHANGES_REQUESTED** [review]

File: src/adapters/AbstractRichtextEditorAdapter.ts (hunk 1 above)
Comment (abhijeetnarvekar) [reply src/adapters/AbstractRichtextEditorAdapter.ts]: Removed

File: src/adapters/AbstractRichtextEditorAdapter.ts (hunk 2 above)
Comment (abhijeetnarvekar) [reply src/adapters/AbstractRichtextEditorAdapter.ts]: Removed

File: src/utils/utils.ts (hunk 3 above)
Comment (abhijeetnarvekar) [reply src/utils/utils.ts]: Removed

File: src/utils/utils.ts (hunk 4 above)
Comment (abhijeetnarvekar) [reply src/utils/utils.ts]: Removed

**Review by pcdeshmukh — APPROVED** [review]

**Pull request merged**